tools/packages/wesl/src/LinkedWesl.ts

```diff
@@ -0,0 +1,171 @@
+export interface WeslGPUCompilationInfo extends GPUCompilationInfo {
+  messages: WeslGPUCompilationMessage[];
+}
+
+export interface WeslGPUCompilationMessage extends GPUCompilationMessage {
+  module: {
+    url: string; // LATER this should be a qualified module path. And something else should map it to a URL that is relative to the correct place.
+    text?: string; // LATER: I don't think that the text should be a part of the compilation message. Instead the module url should be usable as a key.
+  };
+}
+
+/** A {@link GPUValidationError} with an inner error (for a stack trace). Can also point at a WESL source file. */
+export class ExtendedGPUValidationError extends GPUValidationError {
+  public cause?: Error;
+  public compilationInfo?: WeslGPUCompilationInfo;
+  constructor(message: string, options?: { cause?: Error }) {
+    super(message);
+    this.cause = options?.cause;
+  }
+}
+
+/**
+ * Multiple WESL files that have been linked together to produce WGSL code.
+ *
+ * Call {@link LinkedWesl.createShaderModule} on a {@link WeslDevice} to make the error reporting aware of the WESL code.
+ */
+export class LinkedWesl {
+  constructor(public sourceMap: SrcMap) {}
+
+  /**
+   * Creates a {@link GPUShaderModule}.
+   *
+   * The compilation info {@link GPUShaderModule.getCompilationInfo} can be remapped with {@link mapGPUCompilationInfo}
+   * @param device GPUDevice. Preferably a {@link WeslDevice} for better error reporting.
+   * @param descriptor - Description of the {@link GPUShaderModule} to create.
+   */
+  createShaderModule(
+    device: GPUDevice | WeslDevice,
+    descriptor: Omit<GPUShaderModuleDescriptor, "code">,
+  ): GPUShaderModule {
+    // Skip the custom behaviour if we do not have a WESL device.
+    if (!("injectError" in device)) {
+      return device.createShaderModule({
+        ...descriptor,
+        code: this.dest,
+      });
+    }
+
+    device.pushErrorScope("validation"); // Surpress the normal error
+    const module = device.createShaderModule({
+      ...descriptor,
+      code: this.dest,
+    });
+    device.popErrorScope();
+    // And report the error!
+    let { promise, resolve } = Promise.withResolvers<GPUError | null>();
+    device.injectError("validation", promise); // Inject our custom error
+    module.getCompilationInfo().then(compilationInfo => {
+      if (compilationInfo.messages.length === 0) {
+        resolve(null);
+      }
+
+      const mappedCompilationInfo = this.mapGPUCompilationInfo(compilationInfo);
+      const errorMessage = compilationInfoToErrorMessage(
+        mappedCompilationInfo,
+        module,
+      );
+      // Error message cannot be null, since we're passing at least one message to it.
+      assertThat(errorMessage !== null);
+      const error = new ExtendedGPUValidationError(errorMessage, {
+        cause: new Error("createShaderModule failed"),
+      });
+      error.compilationInfo = mappedCompilationInfo;
```

Comment:
We also pass along the compilation info, since we'd like to print super duper pretty errors that point at the source files.